File: app/src/main/java/com/example/android/androidskeletonapp/ui/login/LoginViewModel.java
package com.example.android.androidskeletonapp.ui.login;

import android.util.Patterns;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.android.androidskeletonapp.R;
import com.example.android.androidskeletonapp.data.Sdk;

import org.hisp.dhis.android.core.maintenance.D2Error;
import org.hisp.dhis.android.core.user.User;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class LoginViewModel extends ViewModel {

    private MutableLiveData<LoginFormState> loginFormState = new MutableLiveData<>();
    private MutableLiveData<LoginResult> loginResult = new MutableLiveData<>();

    LoginViewModel() {
    }

    LiveData<LoginFormState> getLoginFormState() {
        return loginFormState;
    }

    LiveData<LoginResult> getLoginResult() {
        return loginResult;
    }

    public Single<User> login(String username, String password, String serverUrl) {
        return Sdk.d2().userModule().logIn(username, password, serverUrl)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSuccess(user -> {
                    if (user != null) {
                        loginResult.postValue(new LoginResult(user));
                    } else {
                        loginResult.postValue(new LoginResult("Login error: no user"));
                    }
                })
                .doOnError(throwable -> {
                    String errorCode = "";
                    try {
                        D2Error d2Error = (D2Error) throwable;
                        errorCode = ": " + d2Error.errorCode();
                    } catch (Exception ignored) {
                    }
                    loginResult.postValue(new LoginResult("Login error" + errorCode));
                    throwable.printStackTrace();
                });
    }

    void loginDataChanged(String serverUrl, String username, String password) {
        if (!isServerUrlValid(serverUrl)) {
            loginFormState.setValue(new LoginFormState(R.string.invalid_server_url, null, null));
        } if (!isUserNameValid(username)) {
            loginFormState.setValue(new LoginFormState(null, R.string.invalid_username, null));
        } else if (!isPasswordValid(password)) {
            loginFormState.setValue(new LoginFormState(null, null, R.string.invalid_password));
        } else {
            loginFormState.setValue(new LoginFormState(true));
        }
    }

    private boolean isServerUrlValid(String serverUrl) {
        if (serverUrl == null) {
            return false;
        }
        return Patterns.WEB_URL.matcher(serverUrl).matches();
    }

    private boolean isUserNameValid(String username) {
        if (username == null) {
            return false;
        }
        return !username.trim().isEmpty();
    }

    private boolean isPasswordValid(String password) {
        return password != null && password.trim().length() > 5;
    }
}

File: app/src/main/java/com/example/android/androidskeletonapp/ui/login/LoginViewModel.kt
package com.example.android.androidskeletonapp.ui.login

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.android.androidskeletonapp.data.Sdk
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.update
import org.hisp.dhis.android.core.maintenance.D2Error

class LoginViewModel internal constructor() : ViewModel() {
    internal val loginResult = MutableLiveData<LoginResult>()
    internal val loginUiState: MutableStateFlow<LoginUiState> = MutableStateFlow(
        LoginUiState(),
    )
    fun login(loginUiState: LoginUiState) {

           Sdk.d2().userModule().logIn(loginUiState.userName!!, loginUiState.password!!, loginUiState.server!!).subscribe(
                { user ->
                    if (user != null) {
                        loginResult.postValue(LoginResult(user))
                    } else {
                        loginResult.postValue(LoginResult("Login error: no user"))
                    }
                },{
                        throwable: Throwable ->
                    var errorCode = ""
                    val d2Error = throwable as D2Error
                    errorCode = ": " + d2Error.errorCode()

                    loginResult.postValue(LoginResult("Login error$errorCode"))
                    throwable.printStackTrace()
                }
            )


    }

    fun setServer(serverUrl: String?) {
        loginUiState.update {
                currentUIState ->
            currentUIState.copy(server = serverUrl)
        }
    }
    fun setUserName(username: String?) {
        loginUiState.update {
                loginUiState ->
            loginUiState.copy(userName = username)
        }
    }

    fun setPassword(username: String?) {
        loginUiState.update {
                loginUiState ->
            loginUiState.copy(userName = username)
        }
    }

    fun initLoginDefaultValues(serverUrl: String?, username: String?, password: String?) {
        loginUiState.update { it.copy(server = serverUrl, userName = username, password = password)  }
    }

}
